### PR TITLE
fix(rust): Add serialization support for newer data types

### PR DESCRIPTION
**Pull request by c-peters**
Add (De)serialization support for Null, Decimal, Array Datatypes

### CODE REVIEW

File: crates/polars-core/src/serde/mod.rs [hunk 1]
```diff
@@ -103,6 +114,36 @@ mod test {
...
     }
 
+    #[test]
+    #[cfg(feature = "dtype-array")]
+    fn test_serde_array_owned_json() {
+        let arrays = vec![
+            UInt8Chunked::new("test", vec![1u8, 2, 3, 4, 5]),
+            UInt8Chunked::new("test2", vec![2u8, 3, 4, 5, 6]),
+        ];
+        let mut builder = get_fixed_size_list_builder(&DataType::UInt8, 2, 5, "array").unwrap();
```
Comment (orlp) [inline crates/polars-core/src/serde/mod.rs:124]: Could you rewrite this test without calling `get_fixed_size_list_builder` and especially without `unsafe` code?

File: crates/polars-core/src/serde/chunked_array.rs [hunk 2]
```diff
@@ -144,7 +159,7 @@ impl Serialize for CategoricalChunked {
         S: Serializer,
     {
         {
-            let mut state = serializer.serialize_map(Some(3))?;
+            let mut state = serializer.serialize_map(Some(4))?;
```
Comment (orlp) [inline crates/polars-core/src/serde/chunked_array.rs:162]: Was this a bug? If so, can you add a failing test case for this? Perhaps we shouldn't only test round-trip but actually the output strings as well.

**Review by orlp — CHANGES_REQUESTED** [review]

**Comment by c-peters** [comment]
Fix #10569

File: crates/polars-core/src/serde/chunked_array.rs [hunk 3]
```diff
@@ -133,6 +134,20 @@ impl_serialize!(Utf8Chunked);
...
+impl_serialize!(ArrayChunked);
+impl Serialize for NullChunked {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_map(Some(3))?;
+        state.serialize_entry("name", self.name())?;
+        state.serialize_entry("datatype", self.dtype())?;
+        state.serialize_entry("values", &self.len())?;
```
Comment (ritchie46) [inline crates/polars-core/src/serde/chunked_array.rs:147]: Shall we call `"values"` -> `"len"`.

File: crates/polars-core/src/serde/series.rs [hunk 4]
```diff
@@ -225,6 +255,17 @@ impl<'de> Deserialize<'de> for Series {
...
                             .unwrap())
                     },
+                    #[cfg(feature = "dtype-array")]
+                    DataType::Array(dt, size) => unsafe {
+                        let mut values: Vec<Series> = map.next_value()?;
+                        let mut builder =
+                            get_fixed_size_list_builder(&dt, values.len(), size, &name).unwrap();
+                        for (i, s) in values.iter_mut().enumerate() {
+                            let s = s.rechunk();
+                            builder.push_unchecked(s.to_arrow(0).as_ref(), i)
```
Comment (ritchie46) [inline crates/polars-core/src/serde/series.rs:265]: Given that we are deserializing here, we also nee the `safe` variants that check the invariants on the builder.

I have only written the unsafe variants now for internal use.

The safe variants can do an invariant check and dispatch to the unsafe ones.

File: crates/polars-core/src/serde/chunked_array.rs (hunk 2 above)
Comment (ritchie46) [reply crates/polars-core/src/serde/chunked_array.rs:162]: This one is already fixed in main.

File: crates/polars-core/src/serde/mod.rs (hunk 1 above)
Comment (ritchie46) [reply crates/polars-core/src/serde/mod.rs:124]: @c-peters an `Array/FixedSizeList` stores all values in a single buffer.

So you can create a `UInt8Array([8, 2, 3, 4, 5, 2, 3, 4, 5, 6])` and store that in the `FixedSizeListArray` (probably under `values`) with the `width` parameter. 


That arrow `FixedSizeListArray` can then be converted into a `Series` with `Series::try_from(("name", array))`.

**Pull request closed without merging**